examples/cuda-dynamic-forall.cpp

```diff
@@ -0,0 +1,150 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-22, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/LICENSE file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include <cstdlib>
+#include <cstring>
+#include <iostream>
+
+#include "memoryManager.hpp"
+
+#include "RAJA/RAJA.hpp"
+
+/*
+ *  Vector Addition Example with resource + dynamic policy selection
+ *
+ *  Computes c = a + b, where a, b, c are vectors of ints using 
+ *  a policy selected at run-time
+ *
+ * If CUDA is enabled, CUDA unified memory is used.
+ */
+
+//
+// Functions for checking and printing results
+//
+void checkResult(int* res, int len);
+void printResult(int* res, int len);
+
+#if defined(RAJA_ENABLE_CUDA)
+using cuda_policy_list = camp::list<RAJA::cuda_exec<256>,
+                                    RAJA::cuda_exec<512>,
+                                    RAJA::cuda_exec<1024>>;
+#endif
+
+
+int main(int argc, char *argv[])
+{
+
+#if defined(RAJA_ENABLE_CUDA)
+
+  if(argc != 2) {
+    RAJA_ABORT_OR_THROW("Usage ./dynamic-forall N");
+  }
+
+  //
+  // Run time policy section is demonstrated in this example by specifying
+  // kernel exection space as a command line argument
+  // Example usage ./dynamic_forall policy N
+  //
+
+  const int pol = std::stoi(argv[1]);
+
+  std::cout << "\n\nRAJA vector addition example...\n";
+  std::cout << "Using policy # "<<pol<<std::endl;
+
+//
+// Define vector length
+//
+  const int N = 1000000;
+
+//
+// Allocate and initialize vector data
+//
+  int *a = memoryManager::allocate<int>(N);
+  int *b = memoryManager::allocate<int>(N);
+  int *c = memoryManager::allocate<int>(N);
+
+  for (int i = 0; i < N; ++i) {
+    a[i] = -i;
+    b[i] = i;
+  }
+
+
+//----------------------------------------------------------------------------//
+
+  std::cout << "\n Running C-style vector addition...\n";
+
+  // _cstyle_vector_add_start
+  for (int i = 0; i < N; ++i) {
+    c[i] = a[i] + b[i];
+  }
+  // _cstyle_vector_add_end
+
+  checkResult(c, N);
+//printResult(c, N);
+
+
+//----------------------------------------------------------------------------//
+// Example of dynamic policy selection for forall
+//----------------------------------------------------------------------------//
+
+  RAJA::resources::Cuda res_gpu;
+
+  RAJA::expt::dynamic_forall<cuda_policy_list>
+  (res_gpu, pol, RAJA::RangeSegment(0, N), [=] RAJA_HOST_DEVICE (int i)   {
```

Comment:
By using a concrete resource we kinda limit the type of policies we can use here. Hopefully this is still okay. I wasn't too sure how to integrate this. @mdavis36 I was wondering if you had some thoughts here.

Reply:
Actually, I have a thought. What if we just supply both host and device resources? Then we can check the policy type and forward the correct resource.

Reply:
We can also consider a typed erased resource like done here: https://github.com/LLNL/RAJA/blob/develop/examples/resource-runtime-teams.cpp#L157 but if I recall correctly a performance hit has been observed.

Reply:
I think a typed erased resource could work here, I don't think the performance hit is that massive. Passing both a host and device would also make sense as I assume the primary use case for using `dynamic_forall` is to be able to swap between host and device execution. However that will feel hacky for use cases when you are switching between policies on a single platform. Alternatively we should be able to do some template magic to require two resources when a policy list contains policies that target more than one platform?

Reply:
Right, I think I like the typed erased resource approach as well. I see what you mean though, sequential and omp (cpu) have different resource types right?

Reply:
I believe sequential and omp cpu would both use 'host' resource.

Reply:
I chatted a bit with @rchen20 about this, using a typed erased resource would not work though because we have to instantiate the resource with every combination of policies the user may provide. So we may end up with mismatches such as: 

```
 error #304: no instance of "forall_impl" matches the argument list
            argument types are: (camp::resources::v1::Cuda, RAJA::policy::simd::simd_exec, const RAJA::RangeSegment, lambda [](int)->void)
```

We could change that into a run-time error, but perhaps not the best thing?

Reply:
I take it back, we can make it work with an erased typed. I turned up the RAJA jamz and it came together. I just pushed it up to the `resource-dynamic-forall` example.